### ISSUE TITLE
hyper dependency is no longer maintained

### ISSUE THREAD

**Issue by kevindice** [Pr0Ger/PyAPNs2#126]
As of January, `hyper` is no longer maintained:
https://github.com/python-hyper/hyper/commit/b77e758f472f00b098481e3aa8651b0808524d84

The readme recommends switching to httpx

**Comment by vitaliylevitskiand** [comment]
also `hyper` requires `h2 <3.0`, which causes problems using other libraries that need h2>=3.0.
I'm facing an issue using `hypercorn`.

**Comment by kevindice** [comment]
My conflict was with Twisted and the http2 extras installed. (Twisted[http2])

**Comment by thierryba** [comment]
Also bother by this...

**Comment by CodeMonk** [comment]
I started looking into what it would take to port from hyper -> httpx, and it's not THAT bad, but, it does take some knowledge of the innerworkings of both packages.

Is anyone doing this? Now knowing hyper or httpx, I'm probably not the best person to do it, but, I do need it done . . . ;-)

Thoughts?

-Dave

**Comment by mcepl** [comment]
Yes, I know, @CodeMonk specifically says https://github.com/CodeMonk/PyAPNs2/commit/a1a3d598a98c84317f60f2f5884760c57a1b8f56 is completely untested and dangerous, but if anybody would have means how to test it, it would be better if the patch was known.

**Comment by CodeMonk** [comment]
I tested it, and declined the PR. Wasn't even close to a full solution. If anyone knows hyper *or* httpx, they could better evaluate how much the lift would be.

I may have some cycles to try to improve my port later.

**Comment by erwan-lemonnier** [comment]
@CodeMonk I have a proper testing environment (staging env with sandboxed apps to play with) and am willing to help, bcz we badly need it. What's the status on your PR? Why did you decline it?

**Comment by erwan-lemonnier** [comment]
FYI, found 2 forks where hyper was replaced with httpx:
https://github.com/pocketlabs/PyAPNs2
https://github.com/karlwnw/PyAPNs2

**Comment by whardeman** [comment]
Good find, @erwan-lemonnier ! Please do share your experience here if you try either of those. I took a look at the first link and the footprint appears to be reasonably small.

**Comment by mcepl** [comment]
> FYI, found 2 forks where hyper was replaced with httpx:
> https://github.com/pocketlabs/PyAPNs2

https://github.com/pocketlabs/PyAPNs2/commit/f1220b8e0b87caf08f3f3d6d3ab9146e75c90543 doesn’t let PyAPNs2 pass its test suite.

> https://github.com/karlwnw/PyAPNs2

**Comment by jayvdb** [comment]
ping everyone , has any of these ports progressed at all?

**Comment by chrigu** [comment]
@jayvdb I just made the parts working that I needed for my project and my port won't be developed any further.

**Comment by erwan-lemonnier** [comment]
And we replaced PyAPNs2 with aioapns, which required a bit of extra work to make sure threads are properly initialised but works really well with python >= 3.8

**Comment by daveisfera** [comment]
Any chance of a release?

**Comment by jayvdb** [comment]
There is some progress at https://github.com/Pr0Ger/PyAPNs2/pull/149

**Comment by daveisfera** [comment]
This is still an issue

**Comment by daveisfera** [comment]
This is still an issue

**Comment by ssyberg** [comment]
Still an issue!

**Comment by salty-horse** [comment]
Don't expect much to happen when you announce it's still an issue on a stale issue of an unmaintained package.

You're welcome to use my PR, which works on modern versions of Python. (I suggest forking it because I can't guarantee my branch will live forever)
https://github.com/Pr0Ger/PyAPNs2/pull/149

**Comment by ssyberg** [comment]
Yea, looks like for anyone having this issue related to django-push-notifications, there's a fix on the verge of merge: https://github.com/jazzband/django-push-notifications/pull/696

**Comment by ajans** [comment]
When migrating to python 3.11, I basically did a self-supported implementation of an apns-client based on `httpx` and certificate-authentication (pem-file) (so please, don't expect any support for this from me if this doesn't work for you).
The only thing I took from this library is the `errors.py` to have a solid error-handling, the same as using this PyAPNS2-library.

https://gist.github.com/ajans/b56353067cc3ad8abbfb35c083ef4a5b

1. This implementation expects a valid module-path in the variable `pem_path` to a certificate-file (pem-file) in your source-tree, adapt if you need other path-types
2. proxy-support: `client.py` looks for the env-variable `https_proxy` and adds the http://-prefix if it is set without that.
3. connections: the httpx-client is configured to only create 1 standing connection without expiry, as the apple-documentation states, this should be enough for most use-cases. If you need more, adapt to your needs.

The method `init_apns_before_retry` is meant for those cases, where e.g. the apple-service resets or discards their side of the connection. In my code using this, I annotate the `_send_push_apple` method with a `tenacity.retry`-annotation to use this init-method before a retry, to reset the connection into working state (1 retry is always enough to reset the connection, but you might adapt to your needs).
The `send_push_apple` method itself should return the status-code `PUSHSTATUS_ERROR_RETRY` if it catches a connection-issue to reset the connection:

```
import <your_package>.apns.client as _apns_client

@_tenacity.retry(
    before_sleep=_apns_client.init_apns_before_retry,
    wait=_tenacity.wait_random(min=1, max=2),
    before=_tenacity.before_log(_log, _logging.WARNING),
    retry=_tenacity.retry_if_result(lambda value: value == PUSHSTATUS_ERROR_RETRY),
    stop=_tenacity.stop_after_attempt(2),
    retry_error_callback=_give_up_retry,
)
def _send_push_apple(
    request,
    push_token,
    apple_payload,
    apple_custom,
    dev,
):
    push_token = push_token_schema.validate(push_token)

    # Send a notification
    status = PUSHSTATUS_OK
    try:
        _log.debug(
            "trying to send apple-push %s to push-token %s",
            str(apple_payload),
            push_token,
        )
        _apns_client.send_notification(push_token, apple_payload, apple_custom, dev)
    # ... don't forget to catch and handle APNS-errors like "BadDeviceToken" and "Unregistered" ...
    except (
        BrokenPipeError,
        _ssl.SSLError,
        _httpx.RequestError,
    ) as exc:
        # another retryable error happened
        status = PUSHSTATUS_ERROR_RETRY
        _log.error(
            "apple-push with %s to push-token %s failed with retryable error %s",
            str(apple_payload),
            push_token,
            str(exc),
        )

    return status


# noinspection PyUnusedLocal
def _give_up_retry(retry_state):  # pylint: disable=unused-argument
    _log.error("This was the last attempt, giving up")
    return PUSHSTATUS_ERROR
```

Maybe this code-snippet might not be easy to integrate into your project, so maybe it would be better to migrate the PyAPNs2-code here or in a fork to use something like this.

But since I don't have the time or energy to do it myself and don't want to constantly ask for support here, I thought it might be good idea to share these snippets with you, since PyAPNS2 helped me a lot before with the initial implementation with older python-versions and I wanted to give something back that might be helpful with these stuck dependencies.

**Comment by JelleZijlstra** [comment]
I started a project https://pypi.org/project/compat-fork-apns2/ aiming to update this project for newer versions of Python. Unfortunately, I don't have enough time and expertise to finish switching away from hyper, but if someone else is interested in contributing code or helping to maintain the fork, please let me know at compat-fork/pyapns2#3.
